vault/identity_store_util.go

```diff
@@ -19,24 +20,57 @@ import (
 	"github.com/hashicorp/vault/logical"
 )
 
+var (
+	errDuplicateIdentityName = errors.New("duplicate identity name")
+)
+
 func (c *Core) loadIdentityStoreArtifacts(ctx context.Context) error {
-	var err error
 	if c.identityStore == nil {
 		c.logger.Warn("identity store is not setup, skipping loading")
 		return nil
 	}
 
-	err = c.identityStore.loadEntities(ctx)
-	if err != nil {
+	loadFunc := func(context.Context) error {
+		err := c.identityStore.loadEntities(ctx)
+		if err != nil {
+			return err
+		}
+		return c.identityStore.loadGroups(ctx)
+	}
+
+	// Load everything when memdb is set to operate on lower cased names
+	err := loadFunc(ctx)
+	switch {
+	case err == nil:
+		// If it succeeds, all is well
+		return nil
+	case err != nil && !errwrap.ContainsType(err, errDuplicateIdentityName):
+		// Return all errors except one
 		return err
 	}
 
-	err = c.identityStore.loadGroups(ctx)
+	// Set identity store to operate on case sensitive identity names
+	c.identityStore.disableLowerCasedNames = true
+
+	// Swap the memdb instance by the one which operates on case sensitive
+	// names, hence obviating the need to unload anything that's already
+	// loaded.
+	err = c.identityStore.resetDB(ctx)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	c.identityStore.logger.Warn("enabling case sensitive identity names")
```

Comment:
Can we move this to L52? It would make the error from `resetDB` above more clear (i.e. error is related to switching over to handling case sensitive names).

Reply:
Done.